framework/math/linear_solver/linear_solver.h

```diff
@@ -78,29 +48,54 @@ class LinearSolver
   void SetKSPSolveSuppressionFlag(bool flag) { suppress_kspsolve_ = flag; }
   bool GetKSPSolveSuppressionFlag() const { return suppress_kspsolve_; }
 
+  /**
+   * Set up the linaer solver
+   */
+  virtual void Setup();
+
+  /**
+   * Solve the system
+   */
+  virtual void Solve();
+
 protected:
+  bool IsSystemSet() const { return system_set_; }
   virtual void PreSetupCallback();
   virtual void SetOptions();
   virtual void SetSolverContext();
   virtual void SetConvergenceTest();
   virtual void SetMonitor();
   virtual void SetPreconditioner();
-
   virtual void SetSystemSize() = 0;
   virtual void SetSystem() = 0;
   virtual void PostSetupCallback();
-
-public:
-  virtual void Setup();
-
-protected:
   virtual void PreSolveCallback();
   virtual void SetInitialGuess() = 0;
   virtual void SetRHS() = 0;
   virtual void PostSolveCallback();
 
-public:
-  virtual void Solve();
+  /// Solver name
+  const std::string solver_name_;
+  const std::string iterative_method_;
+
+  LinSolveContextPtr context_ptr_ = nullptr;
+
+  /// Matrix
+  Mat A_;
+  /// Right-hand side
+  Vec b_;
+  /// Solution vector
+  Vec x_;
+  /// PETSc KSP object
+  KSP ksp_;
```

Comment:
```suggestion
  KSP solver_;
```
```suggestion
  /// PETSc linear solver
 KSP ksp_;
```
Redundant. Consider changing the doc or the member name.

Reply:
I prefer to keep `ksp_`. In PETSc , ksp is how it's called all over the place (example: `KSPSetType(KSP ksp,KSPType method);` ). If someone goes deep enough to see `ksp_`, then they either know what it is or will figure out really quick. I know it goes a little bit against the naming convention, but, to me, a case could be name to ignore the rule a few times, especially if it is a "traditional" name.

Reply:
This is true. The doc saying that `KSP ksp_;` is a KSP object was what I was trying to mitigate. Keeping `ksp_` is fine. It is standard PETSc after all